File: CP/LeetCode/Java/145.java
// 145. Binary Tree Postorder Traversal

import java.util.ArrayList;
import java.util.List;

class Solution {
    static class TreeNode {
        // CSOFF: VisibilityModifier
        int val;
        TreeNode left;
        TreeNode right;

        // CSOFF: VisibilityModifier
        TreeNode(final int x) {
            val = x;
        }
    }

    public List<Integer> postorderTraversal(final TreeNode root) {
        List<Integer> result = new ArrayList<>();
        postorderTraversal(root, result);
        return result;
    }

    private void postorderTraversal(final TreeNode root, final List<Integer> result) {
        if (root == null) {
            return;
        }
        postorderTraversal(root.left, result);
        postorderTraversal(root.right, result);
        result.add(root.val);
    }

    public static void main(final String[] args) {
        Solution solution = new Solution();
        TreeNode root = new TreeNode(1);
        root.right = new TreeNode(2);
        root.right.left = new TreeNode(3);
        System.out.println(solution.postorderTraversal(root));
    }
}

File: CP/LeetCode/C/145.c
// 145. Binary Tree Postorder Traversal

#include <stdio.h>
#include <stdlib.h>

struct TreeNode
{
    int val;
    struct TreeNode *left;
    struct TreeNode *right;
};

int *postorderTraversal(struct TreeNode *root, int *returnSize)
{
    if (root == NULL)
    {
        *returnSize = 0;
        return NULL;
    }

    int *left = NULL;
    int *right = NULL;
    int leftSize = 0;
    int rightSize = 0;

    if (root->left != NULL)
    {
        left = postorderTraversal(root->left, &leftSize);
    }

    if (root->right != NULL)
    {
        right = postorderTraversal(root->right, &rightSize);
    }

    *returnSize = leftSize + rightSize + 1;
    int *result = (int *)malloc(sizeof(int) * (*returnSize));

    for (int i = 0; i < leftSize; i++)
    {
        result[i] = left[i];
    }

    for (int i = 0; i < rightSize; i++)
    {
        result[i + leftSize] = right[i];
    }

    result[leftSize + rightSize] = root->val;

    free(left);
    free(right);

    return result;
}

void free_tree(struct TreeNode *root)
{
    if (root == NULL)
    {
        return;
    }
    free_tree(root->left);
    free_tree(root->right);
    free(root);
}

int main()
{
    struct TreeNode *root = (struct TreeNode *)malloc(sizeof(struct TreeNode));

    root->val = 1;
    root->left = NULL;
    root->right = (struct TreeNode *)malloc(sizeof(struct TreeNode));
    root->right->val = 2;
    root->right->left = (struct TreeNode *)malloc(sizeof(struct TreeNode));
    root->right->left->val = 3;
    root->right->left->left = NULL;
    root->right->left->right = NULL;
    root->right->right = NULL;

    /*
     *       1
     *        \
     *         2
     *        /
     *       3
     */

    int returnSize;
    int *result = postorderTraversal(root, &returnSize);
    for (int i = 0; i < returnSize; i++)
    {
        printf("%d ", result[i]);
    }

    free(result);
    free_tree(root);

    return 0;
}